### ISSUE TITLE
✨ Add filters to request table

### ISSUE THREAD

**Issue by Jerryescandon** [GCTC-NTGC/gc-digital-talent#6365]
## ✨ Feature

Request table has been growing in numbers rapidly, there's a need to start using filters on this table

## 🕵️ Details

Add 2 type of filters to the request table. Filter by:

- Status (multivalue select)
- Department (multivalue select)

## 🎨 Design File

[Link to Figma](https://www.figma.com/file/UsrB15yicgDCLDQmx3snyl/User-Flows?node-id=4716%3A13299&t=zbLzQ0dXanipnzTx-1)

## 📸 Screenshot

![image](https://user-images.githubusercontent.com/16977254/233727947-4e71dfd7-4ff7-4288-b1ff-aa6a585bd3a4.png)

## 🙋‍♀️ Proposed Implementation

(optional) Do you have a proposed implementation?

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] Criteria 1
- [ ] Criteria 2

## 🛑 Blockers

Blocked by {issue numbers}

**Comment by petertgiles** [comment]
Is it correct that both of these filters act as "OR" operators?

`(status = "Done" OR status = "Pending") AND (department = "TBS" OR department = "CRA") `

**Comment by petertgiles** [comment]
@tristan-orourke Should we switch this to a server-driven table with pagination before we add filters to the table?

**Comment by tristan-orourke** [comment]
I don't think we need to switch this to server-driven pagination in order to add the filtering. I've been interested in making a filter modal which would work with frontend-driven filters for a while.

However, if we want to switch this to server pagination anyway, it makes sense to do it first. I guess there could be enough requests that it makes sense to convert it to backend pagination.

**Comment by petertgiles** [comment]
I think we should seriouisly consider it.  This table in prod has 123 pages of data and takes almost 10 seconds to load.
![image.png](https://images.zenhubusercontent.com/6172cf6e898e11fd20507433/b2911a64-6041-439e-8f88-c66757a0223c)

**Comment by Jerryescandon** [comment]
I guess we should also consider a method of removing or archiving old requests, as this will only grow over time.

**Comment by Jerryescandon** [comment]
if we implement filters we could default all statuses on except the "done" and keep it off to improve the first load to only unaddressed requests?

**Comment by vd1992** [comment]
> I think we should seriouisly consider it. This table in prod has 123 pages of data and takes almost 10 seconds to load. 

Concerning!

Can always pass in a value to limit the number returned as a stopgap. Say, pass in limit = 300, given it sorts the data that will be fetched it grabs the newest 300 requests to fill 30 pages max.
Benefit would be that it would be a very, very quick fix whereas refactoring the page would take significantly longer.

![image](https://user-images.githubusercontent.com/40485260/234371276-16a730aa-5083-4127-bcd1-2c8db7897d3f.png)

**Comment by vd1992** [comment]
Not too sure what the best number might be. Maybe 500, or 50 pages. A huge number still but significantly better than 123 pages and rising. 
It would be of course a temporary thing.

**Comment by petertgiles** [comment]
You can see in the screenshot that most of the time is spent waiting for the server.  If we go for a stopgap solution it would need to be something to reduce load on the database, not just reduce the size of the payload.

**Comment by vd1992** [comment]
> You can see in the screenshot that most of the time is spent waiting for the server. If we go for a stopgap solution it would need to be something to reduce load on the database, not just reduce the size of the payload.

Oh, that wouldn't reduce the load? Never mind then.

**Comment by petertgiles** [comment]
> Oh, that wouldn't reduce the load?

Honestly, I don't know.  There are a lot of steps in there:
1. Database query execution
2. Database transmission to webserver
3. Eloquent hydration
4. Lightouse graphql serialization
5. HTTP transmission to client

We know for sure it's not 5.  I'm worried a schema change would reduce it at 4 but I suspect the problem is 1 or 2.  We

**Comment by tristan-orourke** [comment]
I'm pretty sure that the `@limit` directive will actually add a LIMIT clause to the database query, which should improve things at step 1. But honestly, returning 300 results shouldn't be much longer than 10, for a database query. Unless its a _really_ nasty database query.

My suspicion is that the slugishness might be from the authorization step, in the policies. I'll do a bit of testing.

Either way, I'll go ahead with making an issue for converting this to a backend paginated table.

**Comment by tristan-orourke** [comment]
Now blocked by #6389

**Comment by vd1992** [comment]
gave it a whirl, setting limit = 100 made a huge difference for me
seeded 1200 requests, left is all and right is limited

![image](https://user-images.githubusercontent.com/40485260/234392485-c70fc067-ea54-4199-920e-4620b54cc4b1.png)

![image](https://user-images.githubusercontent.com/40485260/234392635-d8fea0f0-8cde-4fc1-bf7b-bf871a0eaccd.png)

**Comment by tristan-orourke** [comment]
I tried logging the database requests which happen during this, and I was wrong, the number of requests happening is not unreasonable. The policy isn't the bottleneck.

I'm curious what the slowest part of this is, but limiting the number of results in #6389 is probably an adequate solution.

**Comment by petertgiles** [comment]
@Jerryescandon Do we need to add the text search as shown in the mockup?  What fields should the text search match?

**Comment by Jerryescandon** [comment]
Nope, sorry that's not in scope for this issue.
